src/infer_request.cc

```diff
@@ -154,7 +154,16 @@ InferenceRequest::SetState(InferenceRequest::State new_state)
       break;
     }
     case InferenceRequest::State::EXECUTING: {
-      if (new_state != InferenceRequest::State::RELEASED) {
+      if (new_state == InferenceRequest::State::RELEASED) {
+        // no-op
+      } else if (new_state == InferenceRequest::State::INITIALIZED) {
+        LOG_WARNING
+            << LogRequest() << " Invalid state transition detected from "
+            << state_ << " to " << new_state
+            << ". When re-using requests, make sure that the request release "
+               "callback is called before calling TRITONSERVER_InferAsync.";
```

Comment:
I'm fine with an interim warning here for now if we're worried about breaking existing cases.

Does it make sense to re-use generate_error here for the first part? Something like:
```cpp
LOG_WARNING << generate_error().Message() << ". When re-using requests ...";
```

Reply:
I think we can relnote this.

> Does it make sense to re-use generate_error here for the first part? Something like:

We would have to delete the error in that case.

Reply:
I think it would be cleaned up automatically because its a `status` not a `TRITONSERVER_Error*`, but I'm not particular to either way. I'm fine with how it is.